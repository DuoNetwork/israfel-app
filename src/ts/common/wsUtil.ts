import WebSocket from 'isomorphic-ws';
import * as CST from '../../../../israfel-relayer/src/common/constants';
import {
	IWsOrderResponse,
	IWsUserOrderResponse
} from '../../../../israfel-relayer/src/common/types';
import Web3Util from '../../../../israfel-relayer/src/utils/Web3Util';
import { IWsAddOrderRequest } from '../common/types';
import util from './util';

class WsUtil {
	public ws: WebSocket | null = null;
	private web3Util: Web3Util | null = null;
	private handleOrderBooksUpdate: ((orderBooks: IWsUserOrderResponse) => any) | null = null;
	public init(host: string) {
		this.web3Util = new Web3Util(window, !__DEV__);
		this.ws = new WebSocket(host);
		this.ws.onopen = function open() {
			console.log('Connected');
		};
		this.ws.onmessage = (m: any) => this.handleMessage(m.data.toString());
	}

	public handleMessage(message: string) {
		const wsMsg: IWsOrderResponse = JSON.parse(message);
		const method = wsMsg.method;
		switch (method) {
			case CST.DB_ADD:
				if (this.handleOrderBooksUpdate) {
					const obRes = wsMsg as IWsUserOrderResponse;
					// const src = others[0];
					if (obRes && obRes.userOrder) {
						// obRes.delay = util.getUTCNowTimestamp() - obRes.timestamp;
						console.log('obRes');
						this.handleOrderBooksUpdate(obRes);
					}
				}
				break;
			default:
				break;
		}
	}

	public subscribe(name: string, marketId: string) {
		if (this.ws && name && marketId) {
			const req = {
				type: 'subscribe',
				channel: {
					name: name,
					marketId: marketId
				},
				requestId: Date.now().toString()
			};
			this.ws.send(JSON.stringify(req));
		}
	}

	public async addOrder(zrxAmt: number, ethAmt: number, isBid: boolean) {
		if (!this.web3Util) throw new Error('error');
		const zrxTokenAddress = this.web3Util.getTokenAddressFromName(CST.TOKEN_ZRX);
		const etherTokenAddress = this.web3Util.getTokenAddressFromName(CST.TOKEN_WETH);
		if (etherTokenAddress === undefined) throw console.error('undefined etherTokenAddress');

		const accounts = await this.web3Util.web3Wrapper.getAvailableAddressesAsync();
		const rawOrder = await this.web3Util.createRawOrder(
			accounts[0],
			__DEV__ ? CST.RELAYER_ADDR_KOVAN : CST.RELAYER_ADDR_MAIN,
			isBid ? zrxTokenAddress : etherTokenAddress,
			isBid ? etherTokenAddress : zrxTokenAddress,
			isBid ? zrxAmt : ethAmt,
			isBid ? ethAmt : zrxAmt,
			Math.ceil(util.getUTCNowTimestamp() / 1000) + 3600
		);
		const pair = 'ZRX-WETH';
		const msg: IWsAddOrderRequest = {
			method: CST.DB_ADD,
			channel: CST.DB_ORDERS,
			pair: pair,
			orderHash: rawOrder.orderHash,
			order: rawOrder.signedOrder
		};
		this.ws.send(JSON.stringify(msg));
	}

	public onOrderBooks(handleOrderBooksUpdate: (orderBooks: IWsUserOrderResponse) => any) {
		this.handleOrderBooksUpdate = handleOrderBooksUpdate;
	}
}

const wsUtil = new WsUtil();
export default wsUtil;
