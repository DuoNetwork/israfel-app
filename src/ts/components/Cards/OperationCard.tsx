import { Select } from 'antd';
import { TimePicker } from 'antd';
import moment, { Moment } from 'moment';
import * as React from 'react';
import * as CST from '../../common/constants';
import wsUtil from '../../common/wsUtil';
import { SDivFlexCenter } from '../_styled';
import { SCard, SCardConversionForm, SCardList, SCardTitle, SInput } from './_styled';

interface IState {
	isCreate: boolean;
	baseCurrency: string;
	description: string;
	price: string;
	targetCurrency: string;
	expireTime: string;
}

export default class OperationCard extends React.PureComponent<{}, IState> {
	constructor(props: object) {
		super(props);
		this.state = {
			isCreate: true,
			baseCurrency: '0',
			description: '',
			targetCurrency: '0',
			price: '0',
			expireTime: '0'
		};
	}

	private handleTypeChange = () =>
		this.setState({
			isCreate: !this.state.isCreate,
			baseCurrency: '0',
			targetCurrency: '0',
			description: '',
			price: '0',
			expireTime: '0'
		});

	private submit = async () => {
		const action = this.state.isCreate ? 'Sell' : 'Buy';
		await wsUtil.addOrder(
			Number(this.state.baseCurrency),
			Number(this.state.targetCurrency),
			action === 'Buy',
			this.state.expireTime
		);
	};

	private getDescription = () => {
		const { isCreate, targetCurrency, baseCurrency } = this.state;
		if (targetCurrency !== '0' && baseCurrency !== '0')
			this.setState({
				price: (Number(targetCurrency) / Number(baseCurrency)).toString()
			});
		else
			this.setState({
				price: '0'
			});

		this.setState({
			description: !isCreate
				? 'Buy ' + baseCurrency + ' with ' + targetCurrency
				: 'Sell ' + baseCurrency + ' with ' + targetCurrency
		});
	};

	private handleAmountInputChange = (value: string) =>
		this.setState({
			baseCurrency: value
		});

	private handlePriceInputChange = (price: string) => {
		const { baseCurrency } = this.state;
		if (price > '0')
			this.setState({
				targetCurrency: (Number(price) * Number(baseCurrency)).toString()
			});
		this.setState({
			price: price
		});
	};

	private handleTargetCurrencyInput = (targetCurrency: string) => {
		this.setState({
			targetCurrency: targetCurrency
		});
	};

	private handleExpireTime(time: Moment, timeString: string) {
		this.setState({
			expireTime: timeString
		});
		console.log(time.toString());
	}

	private handleClear = () =>
		this.setState({
			baseCurrency: '0',
			targetCurrency: '0',
			description: '',
			price: '0'
		});

	public render() {
		const Option = Select.Option;
		const children = [];
		for (let i = 0; i < CST.TH_CURRENCY.length; i++)
			children.push(<Option key={CST.TH_CURRENCY[i]}>{CST.TH_CURRENCY[i]}</Option>);

		return (
			<SCard
				title={<SCardTitle>{CST.TH_ORDER.toUpperCase()}</SCardTitle>}
				width="370px"
				margin="0 0 0 5px"
			>
				<SCardConversionForm>
					<SCardList>
						<div className="status-list-wrapper">
							<ul>
								<li className="block-title">
									<span>{CST.TH_TRADE}</span>
								</li>
								<li>
									<SDivFlexCenter
										horizontal
										width="100%"
										padding="5px 10px 0 0"
										marginBottom="10px"
									>
										<button
											className={
												this.state.isCreate
													? 'conv-button selected'
													: 'conv-button non-select'
											}
											onClick={() => this.handleTypeChange()}
										>
											{CST.TH_SELL}
										</button>
										<button
											className={
												!this.state.isCreate
													? 'conv-button selected'
													: 'conv-button non-select'
											}
											onClick={() => this.handleTypeChange()}
										>
											{CST.TH_BUY}
										</button>
									</SDivFlexCenter>
								</li>
								<li className="input-line bg-dark">
									<Select
										showSearch
										style={{ width: 200 }}
										placeholder={CST.TH_PLACEHOLDER[0]}
										optionFilterProp="children"
										filterOption={(input, option) =>
											option.props.children
												? option.props.children
														.toString()
														.toLowerCase()
														.indexOf(input.toLowerCase()) >= 0
												: null
										}
									>
										{children}
									</Select>
									<SInput
										width="60%"
										className="bg-dark"
										value={this.state.targetCurrency}
										onChange={e =>
											this.handleTargetCurrencyInput(e.target.value)
										}
										onBlur={() => this.getDescription()}
										placeholder="{CST.TT_INPUT_AMOUNT[locale]}"
										right
									/>
								</li>
								<li className={'input-line'}>
									<Select
										showSearch
										style={{ width: 200 }}
										placeholder={CST.TH_PLACEHOLDER[1]}
										optionFilterProp="children"
										filterOption={(input, option) =>
											option.props.children
												? option.props.children
														.toString()
														.toLowerCase()
														.indexOf(input.toLowerCase()) >= 0
												: null
										}
									>
										{children}
									</Select>
									<SInput
										width="60%"
										className={''}
										value={this.state.baseCurrency}
										onChange={e => this.handleAmountInputChange(e.target.value)}
										onBlur={() => this.getDescription()}
										placeholder="{CST.TT_INPUT_AMOUNT[locale]}"
										right
									/>
								</li>
								<li className={'input-line'}>
									<span className="title" style={{ width: 200 }}>
										{CST.TH_PX}
									</span>
									<SInput
										width="60%"
										className={''}
										value={this.state.price}
										onChange={e => this.handlePriceInputChange(e.target.value)}
										onBlur={e => this.handlePriceInputChange(e.target.value)}
										placeholder="{CST.TT_INPUT_AMOUNT[locale]}"
										right
									/>
								</li>
								<li className={'input-line'}>
									<span className="title" style={{ width: 200 }}>
										{CST.TH_EXPIRE.toUpperCase()}
									</span>
									<TimePicker
										onChange={(time: Moment, timeString: string) =>
											this.handleExpireTime(time, timeString)
										}
										defaultValue={moment('8:00:00', 'HH:mm:ss')}
										style={{ width: 170 }}
									/>
								</li>
								<li className="description">
									<div>{this.state.description}</div>
								</li>
								<li>
									<SDivFlexCenter
										horizontal
										width="100%"
										padding="0"
										marginTop="10px"
									>
										<button
											className={'form-button'}
											disabled={
												this.state.baseCurrency === '0' ||
												this.state.targetCurrency === '0'
											}
											onClick={this.submit}
										>
											{CST.TH_SUBMIT}
										</button>

										<button
											className={'form-button'}
											onClick={this.handleClear}
										>
											{CST.TH_CLEAR}
										</button>
									</SDivFlexCenter>
								</li>
							</ul>
						</div>
					</SCardList>
				</SCardConversionForm>
			</SCard>
		);
	}
}
