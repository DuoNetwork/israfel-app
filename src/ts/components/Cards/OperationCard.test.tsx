import { TimePicker } from 'antd';
import { shallow } from 'enzyme';
import * as React from 'react';
import moment = require('../../../../../israfel-relayer/node_modules/moment/moment');
import wsUtil from '../../common/wsUtil';
import { SInput } from './_styled';
import OperationCard from './OperationCard';

describe('OperationCard Test', () => {
	describe('When user search by Price and Volume', () => {
		it('Test Snapshot', () => {
			const wrapper = shallow(<OperationCard />);
			jest.useFakeTimers();
			expect(wrapper).toMatchSnapshot();
		});
		// });

		it('Test SInput Input', async () => {
			const wrapper = shallow(<OperationCard />);
			jest.useFakeTimers();
			await wrapper
				.find(SInput)
				.at(0)
				.simulate('change', {
					target: {
						value: '1'
					}
				});
			await wrapper
				.find(SInput)
				.at(0)
				.simulate('blur');
			await wrapper
				.find(SInput)
				.at(1)
				.simulate('change', {
					target: {
						value: '2'
					}
				});
			await wrapper
				.find(SInput)
				.at(1)
				.simulate('blur');
			expect(wrapper.state('baseCurrency')).toBe('2');
			expect(wrapper.state('targetCurrency')).toBe('1');
			expect(wrapper.state('price')).toBe('0.5');
			setTimeout(() => {
				expect(
					wrapper
						.find('div')
						.at(4)
						.find('li')
						.at(4)
						.find('.description')
				).toBe('Sell 1 with 1');
			}, 1500);
			await wrapper
				.find(SInput)
				.at(2)
				.simulate('change', {
					target: {
						value: '2'
					}
				});
			await wrapper
				.find(SInput)
				.at(1)
				.simulate('blur');
			expect(wrapper.state('baseCurrency')).toBe('2');
			expect(wrapper.state('targetCurrency')).toBe('4');
			expect(wrapper.state('price')).toBe('2');
			setTimeout(() => {
				expect(
					wrapper
						.find('div')
						.at(4)
						.find('li')
						.at(4)
						.find('.description')
				).toBe('Sell 2 with 1');
			}, 100);
			wsUtil.addOrder = jest.fn();
			wrapper
				.find('button')
				.at(2)
				.simulate('click');
			expect(wsUtil.addOrder).toHaveBeenCalled();
		});

		it(' Expire Time Input ', async () => {
			const wrapper = shallow(<OperationCard />);
			jest.useFakeTimers();
			await wrapper
				.find(TimePicker)
				.at(0)
				.simulate('change', {
					target: {
						value: {
							time: moment.now,
							timeString: '08:00:00'
						}
					}
				});
			setTimeout(() => {
				expect(
					wrapper
						.find('div')
						.at(4)
						.find('li')
						.at(4)
						.find('.expireTime')
				).toBe('');
			}, 1500);
		});

		it('Test SInput Input', async () => {
			const wrapper = shallow(<OperationCard />);
			wrapper
				.find('button')
				.at(0)
				.simulate('click');
			jest.useFakeTimers();
			await wrapper
				.find(SInput)
				.at(0)
				.simulate('change', {
					target: {
						value: '1'
					}
				});
			await wrapper
				.find(SInput)
				.at(0)
				.simulate('blur');
			await wrapper
				.find(SInput)
				.at(1)
				.simulate('change', {
					target: {
						value: '2'
					}
				});
			await wrapper
				.find(SInput)
				.at(1)
				.simulate('blur');
			expect(wrapper.state('baseCurrency')).toBe('2');
			expect(wrapper.state('targetCurrency')).toBe('1');
			expect(wrapper.state('price')).toBe('0.5');
			setTimeout(() => {
				expect(
					wrapper
						.find('div')
						.at(4)
						.find('li')
						.at(4)
						.find('.description')
				).toBe('Buy 1 with 2');
			}, 1500);
			await wrapper
				.find(SInput)
				.at(2)
				.simulate('change', {
					target: {
						value: '2'
					}
				});
			await wrapper
				.find(SInput)
				.at(1)
				.simulate('blur');
			expect(wrapper.state('baseCurrency')).toBe('2');
			expect(wrapper.state('targetCurrency')).toBe('4');
			expect(wrapper.state('price')).toBe('2');
			setTimeout(() => {
				expect(
					wrapper
						.find('div')
						.at(4)
						.find('li')
						.at(4)
						.find('.description')
				).toBe('Buy 4 with 2');
			}, 1500);
			await wrapper
				.find(SInput)
				.at(1)
				.simulate('change', { target: { value: '1' } });
			wrapper.find('button').at(1).simulate('click');
			expect(wrapper.state('price')).toBe('0');
			expect(wrapper.state('baseCurrency')).toBe('0');
			expect(wrapper.state('targetCurrency')).toBe('0');
		});

		it('Sell Clear', async () => {
			const wrapper = shallow(<OperationCard />);
			jest.useFakeTimers();
			await wrapper
				.find(SInput)
				.at(1)
				.simulate('change', { target: { value: '1' } });
			wrapper
				.find('button')
				.at(3)
				.simulate('click');
			expect(wrapper.state('baseCurrency')).toBe('0');
			expect(wrapper.state('targetCurrency')).toBe('0');
			expect(wrapper.state('price')).toBe('0');
		});
	});
});

// 		it('Test SInput Input', async () => {
// 			const wrapper = shallow(<OperationCard />);
// 			wrapper
// 				.find('button')
// 				.at(0)
// 				.simulate('click');
// 			jest.useFakeTimers();
// 			await wrapper
// 				.find(SInput)
// 				.at(0)
// 				.simulate('change', { target: { value: '6506.88' } });
// 			await wrapper
// 				.find(SInput)
// 				.at(0)
// 				.simulate('blur');
// 			await wrapper
// 				.find(SInput)
// 				.at(1)
// 				.simulate('change', { target: { value: '6556' } });
// 			await wrapper
// 				.find(SInput)
// 				.at(1)
// 				.simulate('blur');
// 			wrapper
// 				.find('button')
// 				.at(2)
// 				.simulate('click');
// 			expect(wrapper.state('amount')).toBe('6556');
// 			expect(wrapper.state('price')).toBe('6506.88');
// 			setTimeout(() => {
// 				expect(
// 					wrapper
// 						.find('div')
// 						.at(4)
// 						.find('li')
// 						.at(4)
// 						.find('.description')
// 				).toBe('Buy 6556 at price 6506.88');
// 				// expect(submitOrdersMock).toHaveBeenCalled();
// 				expect(wrapper.state('description')).toBe('Buy 6556 at price 6506.88');
// 			}, 1500);
// 		});

// 		it('Sell Clear', async () => {
// 			const wrapper = shallow(<OperationCard />);
// 			jest.useFakeTimers();
// 			await wrapper
// 				.find(SInput)
// 				.at(1)
// 				.simulate('change', { target: { value: '6556' } });
// 			await wrapper
// 				.find('button')
// 				.at(1)
// 				.simulate('click');
// 			expect(wrapper.state('amount')).toBe('0');
// 			expect(wrapper.state('price')).toBe('0');
// 			expect(wrapper.state('description')).toBe('');
// 			setTimeout(() => {
// 				expect(
// 					wrapper
// 						.find('div')
// 						.at(4)
// 						.find('li')
// 						.at(4)
// 						.find('.description')
// 				).toBe('Sell 0 at price 0');
// 			}, 500);
// 		});

// 		it('Buy Clear', async () => {
// 			const wrapper = shallow(<OperationCard />);
// 			wrapper
// 				.find('button')
// 				.at(0)
// 				.simulate('click');
// 			jest.useFakeTimers();
// 			await wrapper
// 				.find(SInput)
// 				.at(1)
// 				.simulate('change', { target: { value: '6556' } });
// 			await wrapper
// 				.find('button')
// 				.at(1)
// 				.simulate('click');
// 			await wrapper
// 				.find('button')
// 				.at(3)
// 				.simulate('click');
// 			wrapper
// 				.find('button')
// 				.at(2)
// 				.simulate('click');
// 			expect(wrapper.state('amount')).toBe('0');
// 			expect(wrapper.state('price')).toBe('0');
// 			expect(wrapper.state('description')).toBe('');
// 			setTimeout(() => {
// 				expect(
// 					wrapper
// 						.find('div')
// 						.at(4)
// 						.find('li')
// 						.at(4)
// 						.find('.description')
// 				).toBe('Buy 0 at price 0');
// 				// expect(submitOrdersMock).toHaveBeenCalled();
// 			}, 500);
// 		});
// 	});
// });
