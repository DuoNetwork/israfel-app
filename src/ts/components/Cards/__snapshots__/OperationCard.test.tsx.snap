// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OperationCard Test When user search by Price and Volume Test Snapshot 1`] = `
<Styled(Card)
  margin="0 0 0 5px"
  title={
    <styled.div>
      OPERATION
    </styled.div>
  }
  width="370px"
>
  <styled.div>
    <styled.div>
      <div
        className="status-list-wrapper"
      >
        <ul>
          <li
            className="block-title"
          >
            <span>
              Trades
            </span>
          </li>
          <li>
            <styled.div
              horizontal={true}
              marginBottom="10px"
              padding="5px 10px 0 0"
              width="100%"
            >
              <button
                className="conv-button selected"
                onClick={[Function]}
              >
                sell
              </button>
              <button
                className="conv-button non-select"
                onClick={[Function]}
              >
                buy
              </button>
            </styled.div>
          </li>
          <li
            className="input-line bg-dark"
          >
            <Select
              choiceTransitionName="zoom"
              filterOption={[Function]}
              optionFilterProp="children"
              placeholder="Target Currency"
              prefixCls="ant-select"
              showSearch={true}
              style={
                Object {
                  "width": 200,
                }
              }
              transitionName="slide-up"
            >
              <Option
                key="ZRX"
              >
                ZRX
              </Option>
              <Option
                key="ETH"
              >
                ETH
              </Option>
            </Select>
            <styled.input
              className="bg-dark"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="{CST.TT_INPUT_AMOUNT[locale]}"
              right={true}
              value="0"
              width="60%"
            />
          </li>
          <li
            className="input-line"
          >
            <Select
              choiceTransitionName="zoom"
              filterOption={[Function]}
              optionFilterProp="children"
              placeholder="Base Currency"
              prefixCls="ant-select"
              showSearch={true}
              style={
                Object {
                  "width": 200,
                }
              }
              transitionName="slide-up"
            >
              <Option
                key="ZRX"
              >
                ZRX
              </Option>
              <Option
                key="ETH"
              >
                ETH
              </Option>
            </Select>
            <styled.input
              className=""
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="{CST.TT_INPUT_AMOUNT[locale]}"
              right={true}
              value="0"
              width="60%"
            />
          </li>
          <li
            className="input-line"
          >
            <span
              className="title"
              style={
                Object {
                  "width": 200,
                }
              }
            >
              Price
            </span>
            <styled.input
              className=""
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="{CST.TT_INPUT_AMOUNT[locale]}"
              right={true}
              value="0"
              width="60%"
            />
          </li>
          <li
            className="input-line"
          >
            <span
              className="title"
              style={
                Object {
                  "width": 200,
                }
              }
            >
              EXPIRE TIME
            </span>
            <TimePicker
              align={
                Object {
                  "offset": Array [
                    0,
                    -2,
                  ],
                }
              }
              defaultValue={"2018-10-31T00:00:00.000Z"}
              disabled={false}
              focusOnOpen={true}
              hideDisabledOptions={false}
              onChange={[Function]}
              placement="bottomLeft"
              prefixCls="ant-time-picker"
              style={
                Object {
                  "width": 170,
                }
              }
              transitionName="slide-up"
            />
          </li>
          <li
            className="description"
          >
            <div />
          </li>
          <li>
            <styled.div
              horizontal={true}
              marginTop="10px"
              padding="0"
              width="100%"
            >
              <button
                className="form-button"
                disabled={true}
                onClick={[Function]}
              >
                Submit
              </button>
              <button
                className="form-button"
                onClick={[Function]}
              >
                Clear
              </button>
            </styled.div>
          </li>
        </ul>
      </div>
    </styled.div>
  </styled.div>
</Styled(Card)>
`;
